libcudacxx/include/cuda/std/__type_traits/is_swappable.h

```diff
@@ -36,22 +37,74 @@
 
 _LIBCUDACXX_BEGIN_NAMESPACE_STD
 
+// We need to detect whether there is already a free function swap that would end up being ambiguous.
+// This can happen when a type pulls in both namespace std and namespace cuda::std via ADL.
+// In that case we are always safe to just not do anything because that type must be host only.
+// However, we must be carefull to ensure that we still create the overload if there is just a hidden friend swap
+namespace __detect_hidden_friend_swap
+{
+// This will intentionally create an ambiguity with std::swap if that is find-able by ADL. But it will not interfere
+// with hidden friend swap
+template <class _Tp>
+void swap(_Tp&, _Tp&);
+
+struct __hidden_friend_swap_found
+{};
+
+template <class _Tp>
+_LIBCUDACXX_INLINE_VISIBILITY auto __swap(_Tp& __lhs, _Tp& __rhs) -> decltype(swap(__lhs, __rhs));
+_LIBCUDACXX_INLINE_VISIBILITY auto __swap(...) -> __hidden_friend_swap_found;
+template <class _Tp>
+struct __has_hidden_friend_swap
+    : is_same<decltype(__detect_hidden_friend_swap::__swap(_CUDA_VSTD::declval<_Tp&>(), _CUDA_VSTD::declval<_Tp&>())),
+              void>
+{};
+} // namespace __detect_hidden_friend_swap
+
+namespace __detect_adl_swap
+{
+struct __no_swap_found
+{};
+template <class _Tp>
+_LIBCUDACXX_INLINE_VISIBILITY auto __swap(_Tp& __lhs, _Tp& __rhs) -> decltype(swap(__lhs, __rhs));
+_LIBCUDACXX_INLINE_VISIBILITY auto __swap(...) -> __no_swap_found;
+template <class _Tp>
+struct __has_no_adl_swap
+    : is_same<decltype(__detect_adl_swap::__swap(_CUDA_VSTD::declval<_Tp&>(), _CUDA_VSTD::declval<_Tp&>())),
+              __no_swap_found>
+{};
+template <class _Tp, size_t _Np>
+struct __has_no_adl_swap_array
+    : is_same<
+        decltype(__detect_adl_swap::__swap(_CUDA_VSTD::declval<_Tp (&)[_Np]>(), _CUDA_VSTD::declval<_Tp (&)[_Np]>())),
+        __no_swap_found>
+{};
+
+// We should only define swap if there is no adl found function or it is a hidden friend
```

Comment:
Nit: You spelled ADL in ALLCAPS in other places.
```suggestion
// We should only define swap if there is no ADL-found function or it is a hidden friend
```